### ISSUE TITLE
[bug] Issues with configuration-specific preprocesor definitions in CMakeToolchain

### ISSUE THREAD

**Issue by mpusz** [conan-io/conan#8284]
First of all, it is not possible to specify just a preprocessor definition without value.

```python
    def generate(self):
        tc = CMakeToolchain(self)
        tc.preprocessor_definitions.debug["gsl_CONFIG_CONTRACT_CHECKING_AUDIT"]
        tc.generate()
```

we have to provide some value:

```python
    def generate(self):
        tc = CMakeToolchain(self)
        tc.preprocessor_definitions.debug["gsl_CONFIG_CONTRACT_CHECKING_AUDIT"] = 1
        tc.generate()
```

this may not work for all the cases as sometimes the preprocessor just checks if a flag is defined or not and does not care about the value.

Next for the above code the following will be generated in a toolchain file:

```cmake
add_definitions(-Dgsl_CONFIG_CONTRACT_CHECKING_AUDIT=$<IF:$<CONFIG:debug>,"1","">)
```

the problem is that the flag is always defined. In order to fix this `add_compile_definitions()` should be used (yeah, I know it was added in CMake 3.12):

```cmake
add_compile_definitions($<$<CONFIG:debug>:gsl_CONFIG_CONTRACT_CHECKING_AUDIT>)
```

or

```cmake
add_compile_definitions($<$<CONFIG:debug>:gsl_CONFIG_CONTRACT_CHECKING_AUDIT=1>)
```

Unfortunatelly, `add_definitions()` seems to not support this syntax.

To summarize:
1. Add support in CMakeToolchain to just specify if a preprocessor flag is present or not (without any values)
2. Fix the toolchain to use `add_compile_definitions()`

**Comment by solarispika** [comment]
Conan 2 looks like already using `add_compile_definitions()` for `tc.preprocessor_definitions`.
https://github.com/conan-io/conan/blob/20517dfb7a6b82f8f8fe3a007ce0c45c775272c4/conan/tools/cmake/toolchain/toolchain.py#L114-L116
However, the first behavior is not yet supported as shown in the code above.

**Comment by memsharded** [comment]
Hi all

I am having a look at this. 
It seems this wouldn't be much of an issue, at the moment it is possible to do something like:
```python
tc.preprocessor_definitions["NOVALUE_DEF"] = None
```
And it will be mapped to ``add_compile_definition(NOVALUE_DEF=None)``, and that works too.

Because having a value should be harmless, such definition would only be used as ``#ifdef``, and thus, giving it any value would be goo to make it work as expected, isn't it? Am I missing something?

**Comment by solarispika** [comment]
Users who need a definition without a value likely do not care what the exact value is. They just want the definition present.

The current design forces those users to arbitrarily come up with a value.

I would say this is more of a UX improvement than a bug. It caters to the full range of compiler use cases.

In summary, while functional, the proposed change would improve ease of use. It handles an additional compiler syntax variation, reducing the configuration burden on users.

BTW, the three major compilers (MSVC, GCC, Clang) default to defining macros as 1 when no value is specified.

I think 1 would be more appropriate than None in this case.

Also, when users convert their build script to use CMakeToolchain, the actual compile arguments might differ as those no-value definitions are now given values.

This might not affect output, but it could invalidate compiler caches like ccache.

**Comment by memsharded** [comment]
The UX improvement should be easy, proposed in https://github.com/conan-io/conan/pull/15545. Could you please check that this is the intended behavior?

**Comment by solarispika** [comment]
Hi Memsharded,
yes, it looks great, thanks!

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/15545 merged for next 2.1
